tracer/src/Datadog.Trace/Logging/DirectSubmission/Sink/PeriodicBatching/BatchingSink.cs

```diff
@@ -0,0 +1,227 @@
+                    {
+                        _queue.TryEnqueue(logEvent);
+                        _started = true;
+
+                        // Special handling to try to get the first event across as quickly
+                        // as possible to show we're alive!
+                        SetTimer(TimeSpan.Zero);
+
+                        return;
+                    }
+                }
+            }
+
+            _queue.TryEnqueue(logEvent);
+        }
+
+        private static void ResetSyncContextAndWait(Func<Task> taskFactory)
+        {
+            var prevContext = SynchronizationContext.Current;
+            SynchronizationContext.SetSynchronizationContext(null);
+            try
+            {
+                taskFactory().Wait();
+            }
+            finally
+            {
+                SynchronizationContext.SetSynchronizationContext(prevContext);
+            }
+        }
+
+        /// <summary>
+        /// Emit a batch of log events, running to completion synchronously.
+        /// </summary>
+        /// <param name="events">The events to emit.</param>
+        protected abstract Task EmitBatch(Queue<DatadogLogEvent> events);
+
+        /// <summary>
+        /// Free resources held by the sink.
+        /// </summary>
+        /// <param name="disposing">If true, called because the object is being disposed; if false,
+        /// the object is being disposed from the finalizer.</param>
+        private void Dispose(bool disposing)
+        {
+            if (!disposing)
+            {
+                return;
+            }
+
+            CloseAndFlush();
+        }
+
+        private async Task OnTick()
+        {
+            try
+            {
+                bool batchWasFull;
+                do
+                {
+                    while (_waitingBatch.Count < _batchSizeLimit &&
+                           _queue.TryDequeue(out var next))
+                    {
+                        _waitingBatch.Enqueue(next);
+                    }
+
+                    if (_waitingBatch.Count == 0)
+                    {
+                        return;
+                    }
+
+                    await EmitBatch(_waitingBatch).ConfigureAwait(false);
+
+                    batchWasFull = _waitingBatch.Count >= _batchSizeLimit;
+                    _waitingBatch.Clear();
+                    _status.MarkSuccess();
+                }
+                while (batchWasFull); // Otherwise, allow the period to elapse
+            }
+            catch (Exception ex)
+            {
+                _logger.Error(ex, "Exception while emitting periodic batch");
+                _status.MarkFailure();
+            }
+            finally
+            {
+                if (_status.ShouldDropBatch)
+                {
+                    _waitingBatch.Clear();
+                }
+
+                if (_status.ShouldDropQueue)
+                {
+                    while (_queue.TryDequeue(out _)) { }
```

Comment:
Why not `.Clear()`?

Reply:
Because BoundedConcurrentQueue doesn't have a clear method, and this was vendored code🙂